core/livepeernode.go

```diff
@@ -51,33 +53,38 @@ type LivepeerNode struct {
 
 	// Transcoder public fields
 	SegmentChans     map[ManifestID]SegmentChan
+	Recipient        pm.Recipient
 	OrchestratorPool net.OrchestratorPool
 	Ipfs             ipfs.IpfsApi
 	ServiceURI       *url.URL
 	OrchSecret       string
 	Transcoder       Transcoder
 
 	// Transcoder private fields
-	segmentMutex *sync.Mutex
-	tcoderMutex  *sync.RWMutex
-	taskMutex    *sync.RWMutex
-	taskChans    map[int64]TranscoderChan
-	taskCount    int64
+	pmSessions      map[ManifestID]map[string]bool
```

Comment:
The full use of `pmSessions` seems to still be a WIP but is there a reason we can't store the sessions locally within the transcode loop rather than keeping it as essentially global state here? The transcode loop is single-threaded which would mitigate the need for locking and auxiliary helpers such as `getAndClearPMSessionIDsByManifestID`. The sessions list could go into the `transcodeConfig` (perhaps with an appropriate rename), or a new struct for mutable state within the transcode loop.

Reply:
It's definitely the initial intent I had, and I'm definitely open to this refactor. One of the challenges I noticed is in the process writing pm sessionIDs to this memory cache. That moment needs to be when we receive a winning ticket - which we will probably unmarshal and process in `ServeSegment`. To store the sessionID then in the transcode loop, I believe we would need to pass the sessionID somehow through some function calls, and eventually add it to the struct we send over the segment channel, in order to get this data point all the way through to `transcodeSegmentLoop`. From a quick look at `SegChanData` it doesn't look like the best place to add payment-related information. Therefore I'm still wondering whether it makes sense to keep this as a global variable such that we can write to it from `ServeSegment` whenever we receive a winning ticket, and then read those sessionIDs in the transcode loop when it's time to redeem.

Would love hear more on how you think we should overcome some of the challenges I mentioned above, and any other challenges you are thinking of!